### ISSUE TITLE
GPU memory leak issue?

### ISSUE THREAD

**Issue by penfe** [flatironinstitute/cufinufft#75]
Dear Yu-hsuan,

I am trying to invoke the cunufft in C++ time and time again, but the monitor for NVIDIA tool [NVTOP](https://github.com/Syllo/nvtop) shows the gpu memory is keep growing.

Would you please kindly review my code below and show me if I have some mistakes in the cunufft API usage? 

~~~c++
#include <cufinufft.h>
#include <complex>
#include <iostream>
#include "unistd.h" // sleep

#define FI_OK 0

void call_finufft(int N1, int N2, int M, int ntransf, float* x, float* y, cuFloatComplex* c, cuFloatComplex* fk)
{
	int maxbatchsize = 0;
	int dim = 2;
	int nmodes[3] = { N1, N2, 1};
	float tol = 1e-6;

	int type = 1;
	int iflag = 1;

	cufinufft_opts opts;
	if (FI_OK != cufinufftf_default_opts(type, dim, &opts))
	{
		std::cout << "err: cufinufft_default_opts" << std::endl;
		return;
	}		
	cufinufftf_plan dplan;
	if (FI_OK != cufinufftf_makeplan(type, dim, nmodes, iflag, ntransf, tol, maxbatchsize, &dplan, NULL))
	{
		std::cout << "err: cufinufftf_makeplan" << std::endl;
		return;
	}
	if (FI_OK != cufinufftf_setpts(M, x, y, NULL, 0, NULL, NULL, NULL, dplan))
	{
		std::cout << "err: cufinufftf_setpts" << std::endl;
		return;
	}
	if (FI_OK != cufinufftf_execute(c, fk, dplan))
	{
		std::cout << "err: cufinufftf_execute" << std::endl;
		return;
	}
	if (FI_OK != cufinufftf_destroy(dplan))
	{
		std::cout << "err: cufinufftf_destroy" << std::endl;
		return;
	}

	std::cout << "ok" << std::endl;
}

void test()
{
	int N1 = 512;
	int N2 = 512;
	int M = 632 * 149;
	int ntransf = 2;

	float *d_x, *d_y;
	CUCPX *d_c, *d_fk;
	cudaMalloc(&d_x,M*sizeof(float));
	cudaMalloc(&d_y,M*sizeof(float));
	cudaMalloc(&d_c,M*ntransf*sizeof(cuFloatComplex));
	cudaMalloc(&d_fk,N1*N2*ntransf*sizeof(cuFloatComplex));

	for (int outer = 0; outer < 30; outer++)
	{
		for (int inner = 0; inner < 5; inner++)
		{
			call_finufft(N1, N2, M, ntransf, d_x, d_y, d_c, d_fk);
		}
		sleep(1);// the gpu memory is keep growing, memory leak?
	}

	cudaFree(d_x);
	cudaFree(d_y);
	cudaFree(d_c);
	cudaFree(d_fk);
}

// test_mem_leak.cu is located at the test folder
// nvcc test_mem_leak.cu -o test_mem_leak -L /usr/local/cuda/lib64 -lcuda -lcufft -I../include -L ../lib -lcufinufft

int main(int argc, char* argv[])
{
	test();
	
	return 0;
}
~~~

**Comment by MelodyShih** [comment]
Hi Penfe, there is some memory issue that was fixed relatively recent (see #74). Would you mind checking if you are using the most updated code? Thanks.

**Comment by penfe** [comment]
Yes Yu-hsuan, I am using the most updated code. 

The #74 report looks like a python only issue, but I have seen the memory leaks in my C++ program (If I didn't mistake the cufinufft APIs call).

Thanks, 
penfe

**Comment by MelodyShih** [comment]
Hi Penfe, sorry for the late reply. I got some time looking into this issue and pushed the fix in #76 . I used `cuda-memcheck --leak-check` and there is no memory leak reported with the updated codes. Could you check whether the changes solve the problem of the growing gpu memory you observed using NVTOP? Thanks.

**Comment by penfe** [comment]
Great fix, my problem is gone with the latest version.
Thanks very much.
